### ISSUE TITLE
iwyu should not think <iostream> is a supplier of anything other than cerr, cin, cout

### ISSUE THREAD

**Issue by ThosRTanner** [include-what-you-use/include-what-you-use#683]
Because `#include <iostream>` adds a lot of complexity to the generated in order to ensure you only have one cerr/cout/cin. However the latest versions seem to consider it's quite OK to have it when all you need is `<ostream>` (or even `<iosfwd>` and in one case `<cstddef>`).

**Comment by kimgr** [comment]
I seem to recall there are mappings from various lower-complexity headers to `<iostream>`, but I was under the impression that it would only be allowed, never suggested.

But of course, if you're not using any features from that monster header, it would be nice to cull it.

**Comment by ThosRTanner** [comment]
I keep finding myself commenting it out to see what breaks and apparently it's being OKd because it includes size_t...

It should be suggested you remove it and replace it with something lest complex.

I think I preferred the old system which was much much picker

**Comment by kimgr** [comment]
> I think I preferred the old system which was much much picker

Which old system is that? I'm not aware of any changes in this area (but I could be forgetting something).

**Comment by ThosRTanner** [comment]
Well, 2 or 3 versions of clang ago, it was definitely was a lot pickier. However, this is the version at work and that recently got updated and I'm not sure which version it's on now. I'll try and find that out.

**Comment by ThosRTanner** [comment]
ah. version is include-what-you-use 0.10 based on clang version 6.0.0 (tags/RELEASE_600/final)

**Comment by ThosRTanner** [comment]
and here's the command line (somewhat obscure settings due to our build setup):

```
include-what-you-use -isystem /opt/rh/devtoolset-6/root/usr/include/c++/6.3.1 -isystem /opt/rh/devtoolset-6/root/usr/include/c++/6.3.1/x86_64-redhat-linux -isystem /opt/rh/devtoolset-6/root/usr/lib/gcc/x86_64-redhat-linux/6.3.1/include test.cc

(test.cc has correct #includes/fwd-decls)
```

test.cc contains
```
#include <iostream>

void fred(std::ostream &s);
```

this should be recommending iosfwd.h

I change test.cc to

```
#include <iostream>

std::size_t fred();
```

and it still says it's OK. with even less justification

**Comment by ThosRTanner** [comment]
And this one is evil:

```
#include <iostream>
#include <sstream>
#include <string>

std::string fred()
{
    std::stringstream s;
    s << "fish";
    return s.str();
}
```

suggests I should remove `sstream`

**Comment by kimgr** [comment]
Yeah, I agree that looks actively harmful. Will have to dig into this more closely.

**Comment by jonesmz** [comment]
I found this issue because IWYU is recommending that I add `#include <iosfwd>` to the following minimal file because of the size_t.

```cpp
#include <string>
std::string foo(size_t size);
```

**Comment by firewave** [comment]
This might have been improved/fixed with 0.21.

**Comment by kimgr** [comment]
@firewave Thanks for noting all these issues -- I collected them under a label `std-stream-headers`.

I've collected all repro cases and tried them locally on the latest IWYU, and somehow it now behaves sanely for all of them. 0.21 is already closed, so I'll investigate some more and see if I can isolate the fix, otherwise I'll just tag them for 0.22 and the public can be let it on the new behavior in the next release.
